### PR TITLE
[mlir][arith][spirv] Convert arith.truncf rounding mode to SPIR-V

### LINKED ISSUE #87050: [mlir][arith][spirv] Convert `arith.truncf` with rounding mode to SPIR-V
#86152 added an optional rounding mode attribute to `arith.truncf`. Due to missing infrastructure in SPIR-V, conversion to the SPIR-V dialect was not implemented as part of that PR.

SPIR-V accepts specifying an operation rounding mode using the `FPRoundingMode` decoration, receiving an [FP rounding mode](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#FP_Rounding_Mode) value. The decoration key is present in the dialect already, but we're missing a representation for the FP rounding mode values. Once we have that, we could convert:

```mlir
func.func @experimental_constrained_fptrunc(%arg0 : f64) {
  %0 = arith.truncf %arg0 tonearesteven : f64 to f32
  %1 = arith.truncf %arg0 downward : f64 to f32
  %2 = arith.truncf %arg0 upward : f64 to f32
  %3 = arith.truncf %arg0 towardzero : f64 to f32
  return
}
```

into:

```mlir
func.func @experimental_constrained_fptrunc(%arg0 : f64) {
  %0 = spirv.FConvertOp %arg0 {FPRoundingMode = #spirv.FPRoundingMode<RTE>} : f64 to f32
  %1 = spirv.FConvertOp %arg0 {FPRoundingMode = #spirv.FPRoundingMode<RTN>} : f64 to f32
  %2 = spirv.FConvertOp %arg0 {FPRoundingMode = #spirv.FPRoundingMode<RTP>} : f64 to f32
  %3 = spirv.FConvertOp %arg0 {FPRoundingMode = #spirv.FPRoundingMode<RTZ>} : f64 to f32
  return
}
```

Note there is no representation of `tonearestaway` in SPIR-V, so conversion should fail in that case.

### PR DESCRIPTION
**Pull request by andfau-amd**
Resolves #87050.

### CODE REVIEW

**Comment by andfau-amd** [comment]
The first commit is from https://github.com/llvm/llvm-project/pull/101546, please review/merge that first.

File: mlir/lib/Conversion/ArithToSPIRV/ArithToSPIRV.cpp [hunk 1]
```diff
@@ -807,6 +807,25 @@ struct TruncIPattern final : public OpConversionPattern<arith::TruncIOp> {
...
+  case arith::RoundingMode::to_nearest_even:
+    return spirv::FPRoundingMode::RTE;
+  case arith::RoundingMode::toward_zero:
+    return spirv::FPRoundingMode::RTZ;
+  case arith::RoundingMode::upward:
+    return spirv::FPRoundingMode::RTP;
+  case arith::RoundingMode::to_nearest_away:
+    // SPIR-V FPRoundingMode decoration has no ties-away-from-zero mode
+    // (as of SPIR-V 1.6)
+    return {};
```
Comment (kuhar) [inline mlir/lib/Conversion/ArithToSPIRV/ArithToSPIRV.cpp:824]:
```suggestion
    return std::nullopt;
```

File: mlir/lib/Conversion/ArithToSPIRV/ArithToSPIRV.cpp [hunk 2]
```diff
@@ -829,15 +848,20 @@ struct TypeCastingOpPattern final : public OpConversionPattern<Op> {
...
-          // TODO: Perform rounding mode attribute conversion and attach to new
-          // operation when defined in the dialect.
-          return failure();
+          if (auto rm =
+                  convertArithRoundingModeToSPIRV(roundingMode.getValue())) {
+            newOp->setAttr(
+                getDecorationString(spirv::Decoration::FPRoundingMode),
+                spirv::FPRoundingModeAttr::get(rewriter.getContext(), *rm));
+          } else {
+            return failure(); // unsupported rounding mode
```
Comment (kuhar) [inline mlir/lib/Conversion/ArithToSPIRV/ArithToSPIRV.cpp:863]: We can emit an error with `rewriter.notifyMatchFailure`

File: mlir/test/Conversion/ArithToSPIRV/arith-to-spirv.mlir [hunk 3]
```diff
@@ -754,6 +754,21 @@ func.func @fptrunc2(%arg0: f32) -> f16 {
...
+  // CHECK: spirv.FConvert %arg0 {fp_rounding_mode = #spirv.fp_rounding_mode<RTE>} : f64 to f32
+  %0 = arith.truncf %arg0 to_nearest_even : f64 to f32
+  // CHECK: spirv.FConvert %arg0 {fp_rounding_mode = #spirv.fp_rounding_mode<RTN>} : f64 to f32
+  %1 = arith.truncf %arg0 downward : f64 to f32
+  // CHECK: spirv.FConvert %arg0 {fp_rounding_mode = #spirv.fp_rounding_mode<RTP>} : f64 to f32
+  %2 = arith.truncf %arg0 upward : f64 to f32
+  // CHECK: spirv.FConvert %arg0 {fp_rounding_mode = #spirv.fp_rounding_mode<RTZ>} : f64 to f32
+  %3 = arith.truncf %arg0 toward_zero : f64 to f32
+  return
+}
```
Comment (kuhar) [inline mlir/test/Conversion/ArithToSPIRV/arith-to-spirv.mlir:769]: Can we add a test case for `to_nearest_away`?

File: mlir/lib/Conversion/ArithToSPIRV/ArithToSPIRV.cpp (hunk 2 above)
Comment (andfau-amd) [reply mlir/lib/Conversion/ArithToSPIRV/ArithToSPIRV.cpp:863]: I tried to do that with:

```cpp
          } else {
            return rewriter.notifyMatchFailure(
                op->getLoc(),
                llvm::formatv("unsupported rounding mode '{0}'", roundingMode));
          }
```

But for some reason, I never see this error message show up in the output, even with `--debug-only=dialect-conversion`. I don't know why that is, but it seems pointless to add this if nobody can see it.

Comment (andfau-amd) [reply mlir/lib/Conversion/ArithToSPIRV/ArithToSPIRV.cpp:863]: Oh, I just discovered that my negative test case fails anyway for another reason. Aha.

File: mlir/lib/Conversion/ArithToSPIRV/ArithToSPIRV.cpp (hunk 1 above)
Comment (andfau-amd) [reply mlir/lib/Conversion/ArithToSPIRV/ArithToSPIRV.cpp]: Done in a4effd454f09906021e467c24f5bc8a85bbece73.

File: mlir/lib/Conversion/ArithToSPIRV/ArithToSPIRV.cpp (hunk 2 above)
Comment (andfau-amd) [reply mlir/lib/Conversion/ArithToSPIRV/ArithToSPIRV.cpp]: Done in a4effd454f09906021e467c24f5bc8a85bbece73.

File: mlir/test/Conversion/ArithToSPIRV/arith-to-spirv.mlir (hunk 3 above)
Comment (andfau-amd) [reply mlir/test/Conversion/ArithToSPIRV/arith-to-spirv.mlir:769]: Done in a4effd454f09906021e467c24f5bc8a85bbece73.

File: mlir/test/Conversion/ArithToSPIRV/arith-to-spirv-unsupported.mlir [hunk 4]
```diff
@@ -1,5 +1,22 @@
 // RUN: mlir-opt -split-input-file -convert-arith-to-spirv -verify-diagnostics %s
 
+///===----------------------------------------------------------------------===//
+// std cast ops
+//===----------------------------------------------------------------------===//
+
+module attributes {
+  spirv.target_env = #spirv.target_env<
```
Comment (kuhar) [inline mlir/test/Conversion/ArithToSPIRV/arith-to-spirv-unsupported.mlir:8]: the `std` dialect was split/deleted a few years ago

```suggestion
// Cast ops
//===----------------------------------------------------------------------===//

module attributes {
  spirv.target_env = #spirv.target_env<
```

**Review by kuhar — APPROVED** [review]
LGTM

**Pull request merged**